### PR TITLE
Add/remove contact group members

### PR DESCRIPTION
**Pull request by shawnadelic**
Add ability to add/remove contact group members to/from groups, using the popup choice widget.

edit_members_final

When ``extra`` is set to 0 on formset, default behavior is to not display an "Add more" button. The "Add more" button copies an existing blank form and appends it to formset, and otherwise we couldn't guarantee a form would be available to copy (which seemed preferable to hard-coding the widget HTML into the JS file).

Refs SHOOP-1981

### CODE REVIEW

**Comment by tulimaki** [comment]
retest this please

File: shoop/core/models/_contacts.py [hunk 1]
```diff
@@ -39,6 +39,12 @@ class Meta:
     def __str__(self):
         return self.safe_translation_getter("name", default="Group<%s>" % (self.identifier or self.id))
 
+    def get_nonmembers(self):
```
Comment (tulimaki) [inline shoop/core/models/_contacts.py]: This does not seem to be used anywhere so I don't think it's needed.

File: shoop/core/models/_contacts.py [hunk 2]
```diff
@@ -39,6 +39,12 @@ class Meta:
...
+        return Contact.objects.exclude(groups=self.id).order_by("name")
+
+    def get_members(self):
```
Comment (tulimaki) [inline shoop/core/models/_contacts.py]: I think this is not needed either. You can use contact_group.members.all().order_by("name") at the formpart.

File: shoop/admin/templates/shoop/admin/contact_groups/edit.jinja [hunk 3]
```diff
@@ -12,15 +12,19 @@
             {% endcall %}
             {% if contact_group.pk %}
                 {% call content_block(_("Members"), "fa-users") %}
-                    {# TODO: This is very bare-bones; should probably have UI to add/remove members etc... #}
-                    <ul>
-                    {% for member in contact_group.members.all() %}
-                        {% set u=shoop_admin.model_url(member, default="") %}
-                        <li>{% if u %}<a href="{{ u }}">{% endif %}{{ member }}{% if u %}</a>{% endif %}</li>
+                    {% if contact_group.members.exists() %}
```
Comment (tulimaki) [inline shoop/admin/templates/shoop/admin/contact_groups/edit.jinja]: I think the "Edit members"-section should just replace this list. Is there a reason you did a whole new view for editing members?

File: shoop/admin/modules/contact_groups/views/members/forms.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            members_to_add = selected_members - current_members
+            members_to_remove = current_members & removed_members
+            for member in members_to_remove:
+                contact_group.members.remove(member)
+            for member in members_to_add:
+                contact_group.members.add(member)
+
+        message_parts = []
+        if members_to_add:
+            message_parts.append(_("New: %d") % len(members_to_add))
```
Comment (tulimaki) [inline shoop/admin/modules/contact_groups/views/members/forms.py]: Better message would be "%(count)s new members added" or something. This new is not descriptive enough IMO.

File: shoop/admin/modules/contact_groups/views/members/forms.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        if members_to_remove:
+            message_parts.append(_("Removed: %d") % len(members_to_remove))
```
Comment (tulimaki) [inline shoop/admin/modules/contact_groups/views/members/forms.py]: Better message would be "%(count)s new members removed".

**Comment by tulimaki** [comment]
Also add tests. At least that new formset should have tests.

File: shoop/admin/modules/contact_groups/views/members/forms.py (hunk 5 above)
Comment (akx) [inline shoop/admin/modules/contact_groups/views/members/forms.py]: Rather `%(count)s members removed`, not new members... :)

**Comment by tulimaki** [comment]
retest this please

**Comment by tulimaki** [comment]
ok to test

**Comment by tulimaki** [comment]
test this please

File: shoop/admin/templates/shoop/admin/contact_groups/edit.jinja [hunk 6]
```diff
@@ -6,23 +6,15 @@
     {% call content_with_sidebar(content_id="contact_group_form") %}
         <form method="post" id="contact_group_form">
             {% csrf_token %}
-            {% call content_block(_("General information"), "fa-info-circle") %}
-                {{ language_dependent_content_tabs(form) }}
-                {{ render_monolingual_fields(form) }}
-            {% endcall %}
+            {% include form.form_defs["base"].template_name %}
```
Comment (tulimaki) [inline shoop/admin/templates/shoop/admin/contact_groups/edit.jinja]: Use ``{% for form_def in form.form_defs.values() instead %}`` then you can use form_def.template_name in include.

File: shoop/admin/templates/shoop/admin/contact_groups/edit.jinja [hunk 7]
```diff
@@ -6,23 +6,15 @@
...
-                    <ul>
-                    {% for member in contact_group.members.all() %}
-                        {% set u=shoop_admin.model_url(member, default="") %}
-                        <li>{% if u %}<a href="{{ u }}">{% endif %}{{ member }}{% if u %}</a>{% endif %}</li>
-                    {% else %}
-                        <li><i>{% trans %}No members.{% endtrans %}</i></li>
-                    {% endfor %}
-                    </ul>
-                {% endcall %}
+                {% include form.form_defs["members"].template_name %}
```
Comment (tulimaki) [inline shoop/admin/templates/shoop/admin/contact_groups/edit.jinja]: After that for change for form defs you can move contact_group.pk check inside members template.

**Comment by tulimaki** [comment]
retest this please

**Comment by suutari-ai** [comment]
test this please

**Pull request merged**